### PR TITLE
feat(webapp): project the planner's route onto the teleop camera

### PR DESCRIPTION
**Pull request by theo-michel**
The mobile controller app draws the agent's planned route on the camera feed; the web cockpit only had it on the map. This ports that overlay to the teleop page as an optional layer.

## What it does

While a navigation is running (navigate-in-sight / mapfree or map navigation), the planner's route is drawn on the video as a green ground ribbon that tapers with distance. A toggle in the right rail — below head tilt — turns it off; the choice persists in `localStorage`. Default on.

## How it works

Ported from `innate-controller-app`'s `src/components/TrajectoryOverlay.tsx`:

- Subscribes to the same `nav_msgs/Path` topics the map widget already uses (`/navigation/plan`, `/mapfree/plan`).
- Re-expresses each plan point relative to the robot's live pose, drops it on the floor plane, and projects it through the head camera's calibrated pinhole model. The intrinsics, the pitch-dependent camera height and `FY_SCALE` are copied from the app so both UIs draw the same line — if the ribbon ever needs re-aiming, `CAMERA` in the new module is the one knob.
- Head pitch comes from `/mars/head/current_position`, same source as the tilt slider.
- Pose composition (AMCL fix + odom delta), the map-vs-odom frame choice from the `Path` header, and the 4-second staleness rule all follow `mapWidget.js`, which already consumes these topics.

Two deliberate differences from mobile, both because the web stage can be any shape:

- Image pixels map through the video's real `object-fit: contain` letterbox rect rather than a stretched container, so alignment holds at any window size.
- The ribbon only draws while the **main** camera is primary — on the arm or top camera a ground projection is meaningless.

Toggling off drops all four subscriptions, not just the drawing. The overlay runs on the simulator too: its main view is the same head camera (`camera_optical_frame` at the driver's FOV), so it gets its own exact camera model — an ideal pinhole from the render's vertical FOV and the URDF's mount height, which rides the head pivot and so does not swing with pitch. Neither the real lens's calibration nor its pitch-height compensation carries over.

## Testing

Verified live in the simulator, whose main view is the same head camera (`camera_optical_frame` at the driver's FOV): the ribbon lies on the floor straight ahead, is centred to the pixel with a clean taper, tracks the floor as the head pitches, stays hidden on the orbit and arm views, and the rail toggle drops and restores it. The projection was also cross-checked against a Three.js camera built the way `sim/viewer` builds it, agreeing to 1e-13 px across a pitch/distance/lateral sweep.

The geometry was developed against throwaway node tests (frame change, clipping, anchoring, taper), which are removed per the repo's testing rule. The rest of `webapp/tests/` still passes and `tsc -p webapp/tsconfig.json` reports no new errors in the touched files.

Not yet verified against a live robot — worth eyeballing whether the ribbon lines up as well as it does on mobile, since the hardware intrinsics are copied constants rather than a fresh calibration.

### CODE REVIEW

File: webapp/js/teleop/trajectoryOverlay.js [hunk 1]
```diff
@@ -0,0 +1,401 @@
...
+  const sp = Math.sin(pitch);
+  const out = [];
+  for (const p of points) {
+    const rotY = -h * cp - p.fwd * sp;
+    const rotZ = -h * sp + p.fwd * cp;
+    if (rotZ <= 0.1) continue;
+    const u = fx * (p.right / rotZ) + cx;
+    const v = fy * (-rotY / rotZ) + cy;
+    if (u < 0 || u > vw || v < 0 || v > vh) continue;
+    out.push({ x: u, y: v, depth: rotZ });
```
[the remark this answers is not in the document]
Comment (theo-michel) [reply webapp/js/teleop/trajectoryOverlay.js]: Addressed in d8d43685. `projectToImage` now returns `segments` — one run per visible stretch — instead of a single flattened list, and the draw loop builds a separate ribbon per segment, so a route that leaves and re-enters the frame draws two disjoint ribbons with no bridge between them. Covered by `a route that leaves the frame and re-enters draws two disjoint ribbons, not a bridge`.

**Comment by axelpey** [comment]
looks good
Screenshot 2026-08-27 at 6 22 03 PM

File: webapp/js/teleop/trajectoryOverlay.js [hunk 2]
```diff
@@ -0,0 +1,418 @@
...
+    }
+    const u = fx * (p.right / rotZ) + cx;
+    const v = fy * (-rotY / rotZ) + cy;
+    if (u < 0 || u > vw || v < 0 || v > vh) {
+      seg = null;
+      continue;
+    }
+    if (!seg) segments.push((seg = []));
+    seg.push({ x: u, y: v, depth: rotZ });
+    if (i === 0 && Math.hypot(p.fwd, p.right) <= ANCHOR_NEAR_M) startAtRobot = true;
```
[the remark this answers is not in the document]
Comment (theo-michel) [reply webapp/js/teleop/trajectoryOverlay.js]: Fixed

**Comment by theo-michel** [comment]
@greptileai

File: webapp/js/teleop/trajectoryOverlay.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (theo-michel) [reply webapp/js/teleop/trajectoryOverlay.js]: Fixed in a2c3481b, then superseded by 04519bd1: the image-space anchor extrapolation is gone entirely — the ribbon now draws only published planner poses, with the video-rect canvas clip trimming it at the frame edge.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: webapp/js/teleop/trajectoryOverlay.js [hunk 3]
```diff
@@ -0,0 +1,416 @@
...
+    seg.push(pt);
+    if (i === 0) firstVisible = true;
+  }
+
+  if (firstVisible && Math.hypot(points[0].fwd, points[0].right) <= ANCHOR_NEAR_M) {
+    const p0 = points[0];
+    const feetZ = -h * sp;
+    const t = feetZ > NEAR_CLIP_M ? 0 : (NEAR_CLIP_M - feetZ) / (p0.fwd * cp);
+    if (t < 1) segments[0].unshift(project(p0.fwd * t, p0.right * t));
+  }
```
[the remark this answers is not in the document]
Comment (theo-michel) [reply webapp/js/teleop/trajectoryOverlay.js]: Fixed in 04519bd1: the prepended connector is removed — every ribbon point is a published pose, the route splits only at the camera near plane, and off-frame poses are trimmed by the canvas clip instead of being culled or synthesized.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: webapp/js/teleop/trajectoryOverlay.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (theo-michel) [reply webapp/js/teleop/trajectoryOverlay.js]: Update: reverted in 94b41497 by author decision — the anchored bottom extrapolation is intended UX (visually verified on a live robot) and stays.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: webapp/js/teleop/trajectoryOverlay.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (theo-michel) [reply webapp/js/teleop/trajectoryOverlay.js]: Update: reverted in 94b41497 by author decision — the feet anchor is intended UX and stays; this finding is acknowledged and accepted as designed.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

**Review by axelpey — APPROVED** [review]

File: webapp/js/teleop/trajectoryOverlay.js [hunk 4]
```diff
@@ -0,0 +1,476 @@
...
+    if (rotZ <= 0.1) {
+      seg = null;
+      continue;
+    }
+    const u = fx * (p.right / rotZ) + cx;
+    const v = fy * (-rotY / rotZ) + cy;
+    if (u < 0 || u > vw || v < 0 || v > vh) {
+      seg = null;
+      continue;
+    }
```
[the remark this answers is not in the document]
Comment (theo-michel) [reply webapp/js/teleop/trajectoryOverlay.js]: Fixed in 42398e66.

`projectToImage` no longer culls poses individually. It walks the route edge by edge and clips each edge to the frame with Liang-Barsky, so a stretch that crosses the view is drawn even when both poses bracketing it fall outside — including the reported case of two consecutive poses off opposite sides.

Depth on a clipped endpoint is interpolated by its reciprocal, since 1/z is what varies linearly along a projected edge, so the ribbon keeps tapering correctly through a clipped stretch.

Only a genuine break still splits a run: leaving the frame, or a pose behind the lens (which has no image position to clip against). Disjoint visible stretches are therefore still never bridged — the existing split-vs-bridge regression still passes, with each run now ending at the frame edge the route actually crossed.

Regression test added: `an edge crossing the view is drawn though both its poses are outside`. Verified it fails against the previous culling loop (which produced zero segments for that input), and confirmed on a live simulator — a sparse two-pose plan straddling the frame now renders a ribbon spanning the full canvas width, clipped exactly to both edges, where it previously drew nothing.

File: webapp/js/teleop/trajectoryOverlay.js [hunk 5]
```diff
@@ -0,0 +1,543 @@
...
+    if (!plan || !vw || !vh || primaryCameraName(session) !== "main") return;
+    const pose = planFrame === "odom" ? odomPose : mapPose();
+    if (!pose) return;
+
+    const camera = video ? REAL_CAMERA : SIM_CAMERA;
+    const { segments, startAtRobot } = projectToImage(robotRelative(plan, pose), pitchDeg, vw, vh, camera);
+    /** @type {Array<Array<{ x: number, y: number }>>} */
+    const polys = [];
+    for (let i = 0; i < segments.length; i++) {
+      const poly = ribbon(segments[i], vw, vh, i === 0 && startAtRobot);
```
[the remark this answers is not in the document]
Comment (theo-michel) [reply webapp/js/teleop/trajectoryOverlay.js:454]: Fixed in a08f389a. Confirmed the scenario reproduced exactly as described.

The flag was decided from the first pose alone, so it survived the loss of the run holding that pose. It is now set where the run is actually opened: on the first edge, `enter === 0` is precisely the condition that the first pose is in frame and heads `segments[0]`. If that edge never survives — the next pose is behind the near plane — no run sets the flag, so the later run that inherits index 0 gets `anchorBottom = false`.

That also subsumes the explicit in-frame bounds test the flag used to duplicate: for a clipped edge, `enter === 0` holds if and only if its start pose is inside the frame.

Regression test added: `a run that only inherits index 0 does not inherit the feet anchor` — first pose visible at 0.38 m, next at 0.05 m (behind the lens), then a far run at 1.0/1.5 m. Verified it fails against the previous rule, which computed `startAtRobot = true` and would have anchored the far run down to the image bottom. The two existing anchor tests (near route anchors, distant and culled-start routes do not) still pass unchanged.
